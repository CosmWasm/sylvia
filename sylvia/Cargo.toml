[package]
name = "sylvia"
version = { workspace = true }
authors = ["Bart≈Çomiej Kuras <bart.k@confio.gmbh>"]
edition = "2021"
description = "Framework for generating trait based CosmWasm smart contracts"
license = "Apache-2.0"
repository = "https://github.com/CosmWasm/sylvia"
homepage = "https://cosmwasm.com"
keywords = ["cosmwasm", "blockchain", "confio"]
categories = ["api-bindings", "wasm"]
readme = "../README.md"

[features]
mt = ["sylvia-derive/mt", "dep:cw-multi-test", "dep:anyhow"]
stargate = ["cosmwasm-std/stargate", "cw-multi-test/stargate"]
iterator = ["cosmwasm-std/iterator"]
cosmwasm_1_1 = ["cosmwasm-std/cosmwasm_1_1", "cw-multi-test/cosmwasm_1_1"]
cosmwasm_1_2 = [
    "cosmwasm-std/cosmwasm_1_2",
    "cw-multi-test/cosmwasm_1_2",
    "sylvia-derive/cosmwasm_1_2",
    "cosmwasm_1_1",
]
cosmwasm_1_3 = [
    "cosmwasm-std/cosmwasm_1_3",
    "cw-multi-test/cosmwasm_1_3",
    "cosmwasm_1_2",
]
cosmwasm_1_4 = [
    "cosmwasm-std/cosmwasm_1_4",
    "cw-multi-test/cosmwasm_1_4",
    "cosmwasm_1_3",
]
cosmwasm_2_0 = [
    "cosmwasm-std/cosmwasm_2_0",
    "cw-multi-test/cosmwasm_2_0",
    "cosmwasm_1_4",
]
cosmwasm_2_1 = [
    "cosmwasm-std/cosmwasm_2_1",
    "cw-multi-test/cosmwasm_2_1",
    "cosmwasm_2_0",
]
cosmwasm_2_2 = ["cosmwasm-std/cosmwasm_2_2", "cosmwasm_2_1"]

[dependencies]
sylvia-derive = { workspace = true }
cosmwasm-std = { workspace = true, features = ["staking"] }
cosmwasm-schema = { workspace = true }
schemars = { workspace = true }
serde = { workspace = true }
serde-cw-value = "0.7.0"
serde-json-wasm = "1.0.1"
konst = "0.3.15"
cw-multi-test = { workspace = true, features = ["staking"], optional = true }
anyhow = { workspace = true, optional = true }
cw-utils = { workspace = true }

[dev-dependencies]
cw-multi-test = { workspace = true, features = ["staking"] }
anyhow = { workspace = true }
cw-storage-plus = { workspace = true }
cw-utils = { workspace = true }
thiserror = { workspace = true }
trybuild = "1.0.101"
itertools = "0.13.0"

[package.metadata.docs.rs]
all-features = true
rustdoc-args = ["--generate-link-to-definition"]
