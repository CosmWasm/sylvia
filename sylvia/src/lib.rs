#![cfg_attr(docsrs, feature(doc_cfg))]
//! Framework for creating CosmWasm Smart Contract with high-level abstraction layer.
//!
//! Most of implementation lies in `sylvia-derive` crate which is reexported here.

pub mod builder;
pub mod ctx;
pub mod into_response;
#[cfg_attr(docsrs, doc(cfg(feature = "mt")))]
#[cfg(feature = "mt")]
pub mod multitest;
pub mod types;
pub mod utils;

#[cfg_attr(docsrs, doc(cfg(feature = "mt")))]
#[cfg(feature = "mt")]
pub use anyhow;
#[cfg_attr(docsrs, doc(cfg(feature = "mt")))]
#[cfg(feature = "mt")]
pub use cw_multi_test;
pub use cw_utils;
pub use sylvia_derive::{contract, entry_points, interface};
pub use {
    cosmwasm_schema as cw_schema, cosmwasm_std as cw_std, schemars, serde,
    serde_cw_value as serde_value,
};
