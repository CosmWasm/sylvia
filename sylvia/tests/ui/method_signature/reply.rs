#![allow(unused_imports, unused_variables)]
use sylvia::ctx::{InstantiateCtx, ReplyCtx};
use sylvia::cw_std::{Addr, Binary, Response, StdResult};

pub mod mismatched_params {
    use super::*;

    pub struct Contract {}

    #[sylvia::contract]
    impl Contract {
        pub const fn new() -> Self {
            Self {}
        }

        #[sv::msg(instantiate)]
        fn instantiate(&self, _ctx: InstantiateCtx) -> StdResult<Response> {
            Ok(Response::new())
        }

        #[sv::msg(reply, handlers=[on_instantiated], reply_on=success)]
        fn first_reply(
            &self,
            _ctx: ReplyCtx,
            #[sv::data(opt, raw)] _data: Option<Binary>,
            param: String,
        ) -> StdResult<Response> {
            Ok(Response::new())
        }

        #[sv::msg(reply, handlers=[on_instantiated], reply_on=error)]
        fn second_reply(&self, _ctx: ReplyCtx, error: String, param: u32) -> StdResult<Response> {
            Ok(Response::new())
        }
    }
}

pub mod mismatched_param_arity {
    use super::*;

    pub struct Contract {}

    #[sylvia::contract]
    impl Contract {
        pub const fn new() -> Self {
            Self {}
        }

        #[sv::msg(instantiate)]
        fn instantiate(&self, _ctx: InstantiateCtx) -> StdResult<Response> {
            Ok(Response::new())
        }

        #[sv::msg(reply, handlers=[on_instantiated], reply_on=success)]
        fn first_reply(
            &self,
            _ctx: ReplyCtx,
            #[sv::data(opt, raw)] _data: Option<Binary>,
            param: String,
        ) -> StdResult<Response> {
            Ok(Response::new())
        }

        #[sv::msg(reply, handlers=[on_instantiated], reply_on=error)]
        fn second_reply(
            &self,
            _ctx: ReplyCtx,
            error: String,
            param: String,
            param: u32,
        ) -> StdResult<Response> {
            Ok(Response::new())
        }
    }
}

pub mod redundant_params {
    use super::*;

    pub struct Contract {}

    #[sylvia::contract]
    impl Contract {
        pub const fn new() -> Self {
            Self {}
        }

        #[sv::msg(instantiate)]
        fn instantiate(&self, _ctx: InstantiateCtx) -> StdResult<Response> {
            Ok(Response::new())
        }

        #[sv::msg(reply, reply_on=success)]
        fn first_reply(
            &self,
            _ctx: ReplyCtx,
            redundant_before1: u32,
            redundant_before2: String,
            #[sv::data(opt, raw)] _data: Option<Binary>,
            #[sv::payload(raw)] param: String,
        ) -> StdResult<Response> {
            Ok(Response::new())
        }

        #[sv::msg(reply, reply_on=success)]
        fn second_reply(
            &self,
            _ctx: ReplyCtx,
            #[sv::data(opt, raw)] _data: Option<Binary>,
            redundant_between1: u32,
            redudnant_between2: String,
            #[sv::payload(raw)] param: String,
            redundant_after1: Binary,
            redundant_after2: Addr,
        ) -> StdResult<Response> {
            Ok(Response::new())
        }

        #[sv::msg(reply, reply_on=success)]
        fn third_reply(
            &self,
            _ctx: ReplyCtx,
            #[sv::data(opt, raw)] _data: Option<Binary>,
            #[sv::payload(raw)] param: String,
            redundant: Binary,
        ) -> StdResult<Response> {
            Ok(Response::new())
        }
    }
}

fn main() {}
