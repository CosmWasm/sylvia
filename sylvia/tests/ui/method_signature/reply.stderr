error: Mismatched parameter in reply handlers.

         = note: Parameters for the `on_instantiated` handler have to be the same.
         = note: Previous parameter defined for the `on_instantiated` handler.

  --> tests/ui/method_signature/reply.rs:26:13
   |
26 |             param: String,
   |             ^^^^^

error: Mismatched quantity of method parameters.

         = note: Both `on_instantiated` handlers should have the same number of parameters.
         = note: Previous definition of on_instantiated handler.

  --> tests/ui/method_signature/reply.rs:55:12
   |
55 |         fn first_reply(
   |            ^^^^^^^^^^^

error: Wrong usage of `#[sv::data]` attribute.

         = note: The `#[sv::data]` attribute can only be used on the first parameter after the `ReplyCtx`.

  --> tests/ui/method_signature/reply.rs:99:35
   |
99 |             #[sv::data(opt, raw)] _data: Option<Binary>,
   |                                   ^^^^^

error: Redundant payload parameter.

         = note: Expected no parameters between the parameter marked with `#[sv::data]` and `#[sv::payload(raw)]`.

  --> tests/ui/method_signature/reply.rs:97:13
   |
97 |             redundant_before1: u32,
   |             ^^^^^^^^^^^^^^^^^

error: Redundant payload parameter.

         = note: Expected no parameters between the parameter marked with `#[sv::data]` and `#[sv::payload(raw)]`.

   --> tests/ui/method_signature/reply.rs:110:13
    |
110 |             redundant_between1: u32,
    |             ^^^^^^^^^^^^^^^^^^

error: Redundant payload parameter.

         = note: Expected no parameters after the parameter marked with `#[sv::payload(raw)]`.

   --> tests/ui/method_signature/reply.rs:125:13
    |
125 |             redundant: Binary,
    |             ^^^^^^^^^
